### PR TITLE
Modify cmake options to work with VS2019

### PR DESCRIPTION
**Pull request by seibert**

### CODE REVIEW

**Comment by seibert** [comment]
checking on build farm: llvmlite_66

**Comment by esc** [comment]
`llvmlite_66` passed 🎉

File: conda-recipes/llvmlite/bld.bat [hunk 1]
```diff
@@ -1,12 +1,22 @@
...
+    @rem VS2017
+    @rem set CMAKE_GENERATOR_ARCH=Win64
+    set CMAKE_GENERATOR_ARCH=x64
 )
-set CMAKE_GENERATOR=Visual Studio 15 2017 %CMAKE_GENERATOR_ARCH%
+@rem for older VS:
+@rem set CMAKE_GENERATOR=Visual Studio 15 2017
+@rem do not set CMAKE_GENERATOR_TOOLKIT
+set CMAKE_GENERATOR=Visual Studio 16 2019
+set CMAKE_GENERATOR_TOOLKIT=v141
```
Comment (stuartarchibald) [inline conda-recipes/llvmlite/bld.bat:19]: Does this now mean llvmlite cannot be built in a conda-recipe with VS 2017?

File: ffi/build.py [hunk 2]
```diff
@@ -45,35 +52,34 @@ def run_llvm_config(llvm_config, args):
...
-            return True  # keep those we don't recognize
-        ver = int(m.group(1))
-        return ver >= 14
-    generators = list(filter(drop_old_vs, generators))
-
-    generators.append('Visual Studio 15 2017' + (' Win64' if is_64bit else ''))
+        arch = os.environ.get("CMAKE_GENERATOR_ARCH", None)
+        toolkit = os.environ.get("CMAKE_GENERATOR_TOOLKIT", None)
+        generators.append(
+            (os.environ.get("CMAKE_GENERATOR"), arch, toolkit)
```
Comment (stuartarchibald) [inline ffi/build.py:69]: This branch is taken as `os.environ.get("CMAKE_GENERATOR")` is true-ish, perhaps store as a var and reuse?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, glad to see this working! Couple of minor things to resolve else good to go.

File: conda-recipes/llvmlite/bld.bat (hunk 1 above)
Comment (seibert) [reply conda-recipes/llvmlite/bld.bat:19]: build.py will always try VS2019 toolkit v141 and then VS2017 if the generator select by the environment variables fails.  (See generators.extend() line later in patch.)

Comment (stuartarchibald) [reply conda-recipes/llvmlite/bld.bat:19]: Found it, thanks.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the fixes, looks good.

**Comment by sklam** [comment]
Two concerns:
- Do we need to sync up the llvmdev recipe to use VS2019 [instead of VS2017](https://github.com/seibert/llvmlite/blob/49fda327d9805690610f8f5230f0c63561e49838/conda-recipes/llvmdev/bld.bat#L7)? 
- Do the llvmlite meta.yaml needs to remove [this line](https://github.com/seibert/llvmlite/blob/49fda327d9805690610f8f5230f0c63561e49838/conda-recipes/llvmlite/meta.yaml#L41)?

**Comment by seibert** [comment]
To the first question, yes.  We should review and merge #570, which makes equivalent fixes to the llvmdev script for VS2019.  (It has to do fancy things in cmd.exe since there is no equivalent to build.py to put the toolkit searching logic into.)

As for the second question, I don't know.  There is no conda package for anything after `vs2015_runtime`, so I assume it continues to work for later VS releases.  I'm assuming things are OK since all the tests pass.

**Comment by seibert** [comment]
note that in terms of binary compatibility, this PR is continuing to use the same compilers as VS2017 on VS2019 by setting the toolkit, so it should not matter whether we build with VS2017 or VS2019.

**Comment by esc** [comment]
Running again at: `llvmlite_win_only_03`.

**Comment by esc** [comment]
`llvmlite_win_only_03` passed.

**Comment by sklam** [comment]
> this PR is continuing to use the same compilers as VS2017 on VS2019 by setting the toolkit, so it should not matter whether we build with VS2017 or VS2019.

I have memory of how 141 compiler on 2019 and 2017 are not binary compatible. That's why I stopped updating everything to VS2019 back then.

**Comment by esc** [comment]
Should we perhaps make a package from these changes and then give it a spin through the Numba test-suite?

**Comment by sklam** [comment]
Yes, we should

**Review by esc — APPROVED** [review]
Tested on the farm, builds fine on new Windows workers.

**Pull request merged**